### PR TITLE
Fix missing enums and fix MultiDrawElementsIndirectCount parameter type.

### LINKED ISSUE #752: SizedInternalFormat values missing
In the latest stable version in OpenGL4 namespace SizedInternalFormat all enum values for RGB* are missing.

* Windows10
* Native backend
* NuGet packet 3.0.0
* .NET Framework 4.7.1

### LINKED ISSUE #993: Missing GL_TEXTURE_CUBE_MAP_ARRAY in TextureTarget3d enum
### Description

Enum `OpenTK.Graphics.OpenGL4.TextureTarget3d` does not contain a definition for `TextureCubeMapArray`.

According to [wiki](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glTexStorage3D.xhtml) the `glTexStorage3D` (`OpenTK.Graphics.OpenGL4.GL.TexStorage3D`) function's one possible target is `GL_TEXTURE_CUBE_MAP_ARRAY`. However `GL.TexStorage3D` target is `TextureTarget3d` enum and it has no member similar to `GL_TEXTURE_CUBE_MAP_ARRAY`.

### Example

I'm trying to create CubeMap array texture this way:
```c#
using System;
using System.Drawing;
using System.Drawing.Imaging;
using OpenTK.Graphics.OpenGL4;
using PixelFormat = System.Drawing.Imaging.PixelFormat;
using GL_PixelFormat = OpenTK.Graphics.OpenGL4.PixelFormat;
...

public int CreateCubemapArray(string[] imagePaths, int count)
{
    // Generate an array texture
    int texture = GL.GenTexture();
    GL.BindTexture(TextureTarget.TextureCubeMapArray, texture);
    GL.TexStorage3D(
        TextureTarget3d.TextureCubeMapArray, // does not work
        9, 
        SizedInternalFormat.Rgba8, 
        512, 512, 
        count);

    // Load images
    for (int i = 0; i < count * 6; i++)
    {
        LoadImage(imagePaths[i], 512, 512, TextureTarget.TextureCubeMapPositiveX + (i % 6), i / 6);
    }

    // Setting behavior
    GL.TexParameter(TextureTarget.TextureCubeMapArray, TextureParameterName.TextureMinFilter, (int)TextureMinFilter.Linear);
    GL.TexParameter(TextureTarget.TextureCubeMapArray, TextureParameterName.TextureMagFilter, (int)TextureMagFilter.Nearest);

    GL.TexParameter(TextureTarget.TextureCubeMapArray, TextureParameterName.TextureWrapS, (int)TextureWrapMode.ClampToEdge);
    GL.TexParameter(TextureTarget.TextureCubeMapArray, TextureParameterName.TextureWrapT, (int)TextureWrapMode.ClampToEdge);
    GL.TexParameter(TextureTarget.TextureCubeMapArray, TextureParameterName.TextureWrapR, (int)TextureWrapMode.ClampToEdge);

    // Generate Mipmaps
    GL.GenerateMipmap(GenerateMipmapTarget.TextureCubeMapArray);

    return texture;
}

void LoadImage(string path, int width, int height, TextureTarget textureTarget, int depth)
{
    using Bitmap image = new Bitmap(path);
    BitmapData imageData = image.LockBits(
        new Rectangle(0, 0, image.Width, image.Height),
        ImageLockMode.ReadOnly,
        PixelFormat.Format32bppArgb);
    GL.TexSubImage3D(
        textureTarget,
        0, // level
        0, 0, 0, // offsets
        width, height,
        depth,
        GL_PixelFormat.Bgra,
        PixelType.UnsignedByte,
        imageData.Scan0);
    image.UnlockBits(imageData);
}
```

### Related information 

* OpenTK 3.1.0
* .NET Core 3.0

### LINKED ISSUE #1319: Version 3.3.2 missing enums for opengl 4.6 GL_ARB_indirect_parameters
### Description
There appears to be missing enum identifiers in enum.BufferTarget and enum.GetIndexedPName for the following 4.6 additional parameter buffer.

From https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_indirect_parameters.txt:

New Tokens

    Accepted by the <target> parameters of BindBuffer, BufferData,
    BufferSubData, MapBuffer, UnmapBuffer, GetBufferSubData,
    GetBufferPointerv, MapBufferRange, FlushMappedBufferRange,
    GetBufferParameteriv, and CopyBufferSubData:

        PARAMETER_BUFFER_ARB                                    0x80EE

    Accepted by the <value> parameter of GetIntegerv, GetBooleanv, GetFloatv,
    and GetDoublev:

        PARAMETER_BUFFER_BINDING_ARB                            0x80EF

### Repro steps

1. Try defined GL.BindBuffer(BufferTarget.ParameterBuffer, 0) - it fails
2. Missing   GL.GetInteger(GetIndexedPName.ParameterBufferBinding)

### Expected behavior

GL.BindBuffer(BufferTarget.ParameterBuffer, bufid) etc

Please provide a description of the behavior you expect.

### Related information 
OpenGL 3.3.2  .Net NUGET package.

* Workarounds (if any)
   GL.BindBuffer((BufferTarget)0x80ee, Id); - this works, and i've demonstrated it.

### PR DESCRIPTION
**Pull request by NogginBops**
### Purpose of this PR

Fixes most (if not all) of the missing enum entries mentioned in #1062

Fixes #752
Fixes #993
Fixes #1319

This PR could also be backported to OpenTK 3 quite easily by cherry-picking these commits.

### Testing status

Built locally and the new enums are available and working.

### CODE REVIEW

**Comment by BoyBaykiller** [comment]
I think since `TEXTURE_MAX_ANISOTROPY` was added to `TextureParameterName` we should also add it to `GetTextureParameter` below here
https://github.com/opentk/opentk/blob/43754f4a8ad997b65b501e109acbc37b628ff3be/src/Generator.Bind/Specifications/GL2/overrides.xml#L3650

**Comment by NogginBops** [comment]
@BoyBaykiller good catch! Just added it 🙂

**Comment by BoyBaykiller** [comment]
I am not sure sure if I can commit to this PR so I am just gonna mention:
`TEXTURE_CUBE_MAP_SEAMLESS` is an other enum value that could be added.
Its part of `ARB_seamless_cubemap_per_texture` and `AMD_seamless_cubemap_per_texture`

File: src/Generator.Bind/GL2/Generator.cs [hunk 1]
```diff
@@ -120,11 +120,33 @@ public virtual void Process()
...
+
+            foreach(var (name, @enum) in Enums)
+            {
+                var match = Utilities.Extensions.Match(name);
+                if (match.Index == 0 && match.Length != 0)
+                {
+                    extensions.Add(name);
+                }
+            }
+
```
Comment (seanofw) [inline src/Generator.Bind/GL2/Generator.cs:144]: Writing this section as Linq would likely make it a lot more obvious that these loops are doing nothing more than populating the HashSet with a projection of the original data:
```cs
extensions.UnionWith(Delegates
    .SelectMany(d => (d.Item2[0].Category ?? string.Empty).Split('|'))
    .Where(s => !string.IsNullOrEmpty(s));

Match match;
extensions.UnionWith(Enums
    .Select(e => e.Item1)
    .Where(s => (match = Utilities.Extensions.Match(s)).Index == 0 && match.Length != 0));
```
By writing it as Linq, it's short, simple, and a lot more obvious that it's just projecting and filtering, and importantly, it's more obvious that there are no side effects in the loops.

File: src/Generator.Bind/Structures/Constant.cs [hunk 2]
```diff
@@ -72,6 +72,8 @@ public string Value
         /// </summary>
         public string Reference { get; set; }
 
+        public string AdddedInExtension { get; set; }
```
Comment (seanofw) [inline src/Generator.Bind/Structures/Constant.cs:75]: Spelling error here:  `Addded`.

**Comment by seanofw** [comment]
This PR feels like it's two (or more?) PRs combined together; would it make sense to split it up?  I see generator stuff for the enums, mouse stuff that doesn't seem to be related to that, and other stuff for MultiDrawElementsIndirectCount, and I'm having a hard time seeing how all those are really related to each other.

File: src/Generator.Bind/CSharpSpecWriter.cs [hunk 3]
```diff
@@ -410,7 +410,29 @@ private void WriteConstants(BindStreamWriter sw, IEnumerable<Constant> constants
...
                     sw.WriteLine("/// <summary>");
-                    sw.WriteLine("/// Original was " + Settings.ConstantPrefix + c.OriginalName + " = " + c.Value);
+
+                    string requires_string = "";
+                    if (string.IsNullOrEmpty(c.AdddedInExtension) == false)
+                    {
+                        requires_string = "[requires: ";
+
+                        // This constant was added in an extension
+                        if (c.AdddedInExtension.StartsWith("VERSION") && c.AdddedInExtension.Length == "VERSION_X_X".Length)
```
Comment (seanofw) [inline src/Generator.Bind/CSharpSpecWriter.cs:420]: I had to read this several times to figure out what I was looking at.  Maybe consider a regex with named captures here instead, and a ternary to reduce nesting:
```cs
private static readonly Regex _versionRegex = new Regex(@"^VERSION_(?<Major>\d+)_(?<Minor>\d+)$");

...

string version = ((match = _versionRegex.Match(c.AdddedInExtension)).Success)
    ? $"v{match.Groups("Major")}.{match.Groups("Minor")}"
    : c.AdddedInExtension;

requires_string = $"[requires: {version}]";
```
This is also more flexible than the original; it's far easier to tweak the regex to describe the pattern than to modify `StartsWith()` and length comparisons and character hacking.

Also:  Wouldn't the correct name be `requiresString`?

File: src/Generator.Bind/GL2/Generator.cs (hunk 1 above)
Comment (NogginBops) [reply src/Generator.Bind/GL2/Generator.cs:144]: Making this into linq would not improve readability in my mind. I will keep it like it is.

File: src/Generator.Bind/EnumProcessor.cs [hunk 4]
```diff
@@ -56,15 +56,15 @@ public EnumProcessor(IBind generator, IEnumerable<string> overrides)
             Overrides = overrides;
         }
 
-        public EnumCollection Process(EnumCollection enums, string apiname)
+        public EnumCollection Process(EnumCollection enums, string apiname, HashSet<string> extensions)
```
[the remark this answers is not in the document]
Comment (NogginBops) [reply src/Generator.Bind/EnumProcessor.cs:59]: This is a fair comment, though I don't think it's that big of a concern in this code. This code already has a style of very mutable state, and doing anything that doesn't match the current style would draw a lot of attention to it as special for some reason.
So making this `IReadOnlyCollection<string>` would make it safer for modification, but it would make the readability of the code harder as one would wonder why it's a `IReadOnlyCollection<string>` and not just something like `List<string>`.

Basically I think it's not worth it to change that for this code that is basically at the end of it's maintenance. We are going to retire this code with OpenTK 5.

[the remark this answers is not in the document]
Comment (seanofw) [reply src/Generator.Bind/EnumProcessor.cs:59]: Then consider `IReadOnlySet<string>`.  When in doubt, prefer to pass around immutable data; it makes the _correctness_ of the code far easier to demonstrate.  Knowing that `Contains()` runs in O(1) time is useful, but being certain that a callee can't modify the data allows everything that works with that data to be more easily refactored and modified in the future.  Slow code that works with immutable data can readily be made faster; but fast code that works with mutable data can often be hard or even impossible to modify, no matter how good it is.

**Comment by NogginBops** [comment]
This is definitely a PR that touches different things. The GLFW thing was adding a missing enum mentioned in #1062 so as this PR was addressing that issue I decided to put that change into this PR too.

While adding enums from an extension to an existing enum group I noticed that enums didn't have their "origin" in their comments, which would cause a user to maybe mistakenly think a particular enum was available to use in a function without having a specific extension present. So I decided to give adding enum "origin" a go, and that is why it's also included into this PR.

File: src/Generator.Bind/EnumProcessor.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (seanofw) [reply src/Generator.Bind/EnumProcessor.cs:59]: And even if this code is being retired for OpenTK 5, that still does not justify writing it less maintainably.  Even after OpenTK 5 is released, OpenTK 4 will still exist, and we may even need to fix bugs in it and release patched versions.  We owe it to our future selves to make this as maintainable as possible.

File: src/Generator.Bind/GL2/Generator.cs (hunk 1 above)
Comment (NogginBops) [reply src/Generator.Bind/GL2/Generator.cs:144]: I understand your concern, but I do not share it. This code is simple to look at, simple to verify.
The code is already full of mutable state that changes in much more cursed ways than potentially allowing this one hash set to be modified.

**Comment by NogginBops** [comment]
I'm going to merge this now.
The requires strings for enums is a good addition I feel as it's nice to have a reminder that some features (enums) might not be available for the version targeted.

This PR contains a slight breaking feature of changing the one of the parameter types of `GL.MultiDrawElementsIndirectCount` from `All` to `DrawElementsType`. This is deemed acceptable as this is expected to be a relatively rarely used function, and fixing the resulting error should be as easy as a cast to `DrawElementsType` (or just using `DrawElementsType` directly).

Being able to fix all of these enum issues is going to greatly improve the ergonomics of using OpenTK 4.

**Comment by NogginBops** [comment]
Regarding the breaking change on `MultiDrawElementsIndirectCount`, I've decided that I'm going to add an overload for that function which retains the old signature. That way this will no longer be a breaking change.
